compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -601,6 +600,29 @@ trait Checking {
   /** Verify classes extending AnyVal meet the requirements */
   def checkDerivedValueClass(clazz: Symbol, stats: List[Tree])(implicit ctx: Context) =
     Checking.checkDerivedValueClass(clazz, stats)
+
+  /** Given a parent `parent` of a class `cls`, if `parent` is a trait check that
+   *  the superclass of `cls` derived from the superclass of `parent`.
+   *
+   *  An exception is made if `cls` extends `Any`, and `parent` is a Java class
+   *  that extends `Object`. For instance, we accept code like
+   *
+   *      ... extends Any with java.io.Serializable
+   *
+   *  The standard library relies on this idiom.
+   */
+  def checkTraitInheritance(parent: Symbol, cls: ClassSymbol, pos: Position)(implicit ctx: Context): Unit = {
+    parent match {
+      case parent: ClassSymbol if parent is Trait =>
+        val psuper = parent.superClass
+        val csuper = cls.superClass
+        val ok = csuper.derivesFrom(psuper) ||
+          parent.is(JavaDefined) && csuper == defn.AnyClass && psuper == defn.ObjectClass
```

Comment:
I think this might be too permissive: as far as I can tell, `scalac` only allows allows this for two Java classes: `Serializable` and `Comparable` which are treated as if they were traits that extend `Any`: https://github.com/scala/scala/blob/cac6383e6658dc5956540b76b9b46c3b664774ac/src/reflect/scala/reflect/internal/Definitions.scala#L365-L366 If we allow any Java class to be used in universal traits, then we break the definition of universal trait.